hbase-server/src/main/java/org/apache/hadoop/hbase/security/provider/GssSaslServerAuthenticationProvider.java

```diff
@@ -0,0 +1,118 @@
+package org.apache.hadoop.hbase.security.provider;
+
+import java.io.IOException;
+import java.security.PrivilegedExceptionAction;
+import java.util.Map;
+
+import javax.security.auth.callback.Callback;
+import javax.security.auth.callback.CallbackHandler;
+import javax.security.auth.callback.UnsupportedCallbackException;
+import javax.security.sasl.AuthorizeCallback;
+import javax.security.sasl.Sasl;
+import javax.security.sasl.SaslException;
+
+import org.apache.hadoop.hbase.HBaseInterfaceAudience;
+import org.apache.hadoop.hbase.security.AccessDeniedException;
+import org.apache.hadoop.hbase.security.SaslUtil;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.hadoop.security.token.SecretManager;
+import org.apache.hadoop.security.token.TokenIdentifier;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.apache.yetus.audience.InterfaceStability;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@InterfaceAudience.LimitedPrivate(HBaseInterfaceAudience.AUTHENTICATION)
+@InterfaceStability.Evolving
+public class GssSaslServerAuthenticationProvider extends GssSaslAuthenticationProvider
+    implements SaslServerAuthenticationProvider {
+  private static final Logger LOG = LoggerFactory.getLogger(
+      GssSaslServerAuthenticationProvider.class);
+
+  @Override
+  public AttemptingUserProvidingSaslServer createServer(
+      SecretManager<TokenIdentifier> secretManager,
+      Map<String, String> saslProps) throws IOException {
+    UserGroupInformation current = UserGroupInformation.getCurrentUser();
+    String fullName = current.getUserName();
+    LOG.debug("Server's Kerberos principal name is {}", fullName);
+    String[] names = SaslUtil.splitKerberosName(fullName);
+    if (names.length != 3) {
+      throw new AccessDeniedException(
+          "Kerberos principal does NOT contain an instance (hostname): " + fullName);
+    }
+    try {
+      return current.doAs(new PrivilegedExceptionAction<AttemptingUserProvidingSaslServer>() {
+        @Override
+        public AttemptingUserProvidingSaslServer run() throws SaslException {
+          return new AttemptingUserProvidingSaslServer(Sasl.createSaslServer(
+              getSaslAuthMethod().getSaslMechanism(), names[0], names[1], saslProps,
+              new SaslGssCallbackHandler()), () -> null);
+        }
+      });
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException("Failed to construct GSS SASL server");
+    }
+  }
+
+  /** CallbackHandler for SASL GSSAPI Kerberos mechanism */
+  private static class SaslGssCallbackHandler implements CallbackHandler {
+
+    /** {@inheritDoc} */
+    @Override
+    public void handle(Callback[] callbacks) throws UnsupportedCallbackException {
+      AuthorizeCallback ac = null;
+      for (Callback callback : callbacks) {
+        if (callback instanceof AuthorizeCallback) {
+          ac = (AuthorizeCallback) callback;
+        } else {
```

Comment:
Here should have realm callback for cross-realm authentication? 
We can implement it later, just to raise it.

Reply:
I'm not sure enough on the GSSAPI implementation in SASL. We've always done cross-realm trust external to HBase (e.g. set up cross-realm trust for your KDC). I want to leave this one as it is, rather than open some security issue changing stuff I don't understand fully :)